62.CAD/main.cpp

```diff
@@ -234,52 +407,866 @@ class CPolyline
+}
+
+double Hatch::QuadraticBezier::intersectOrtho(double coordinate, int major) const
+{
+	// https://pomax.github.io/bezierinfo/#intersections
+	double points[3];
+	for (uint32_t i = 0; i < 3; i++)
+		points[i] = major ? p[i].y : p[i].x;
+
+	for (uint32_t i = 0; i < 3; i++)
+		points[i] -= coordinate;
+
+	float64_t2 roots = hatchutils::getCurveRoot(points[0], points[1], points[2]);
+	if (roots.x >= 0.0 && roots.x <= 1.0) return roots.x;
+	if (roots.y >= 0.0 && roots.y <= 1.0) return roots.y;
+	return core::nan<double>();
+}
+
+float64_t2 Hatch::QuadraticBezier::evaluateBezier(double t) const
+{
+	float64_t2 position = p[0] * (1.0 - t) * (1.0 - t)
+		+ 2.0 * p[1] * (1.0 - t) * t
+		+ p[2] * t * t;
+	return position;
+}
+
+// https://pomax.github.io/bezierinfo/#pointvectors
+float64_t2 Hatch::QuadraticBezier::tangent(double t) const
+{
+	// TODO: figure out a tangent algorithm for when this becomes A, B, C
+	auto derivativeOrder1First = 2.0 * (p[1] - p[0]);
+	auto derivativeOrder1Second = 2.0 * (p[2] - p[1]);
+	auto tangent = (1.0 - t) * derivativeOrder1First + t * derivativeOrder1Second;
+	auto len = sqrt(tangent.x * tangent.x + tangent.y * tangent.y);
+	return tangent / len;
+}
+
+// https://pomax.github.io/bezierinfo/#extremities
+std::array<double, 4> Hatch::QuadraticBezier::getRoots() const
+{
+	// Quadratic coefficients
+	float64_t2 A = p[0] - 2.0 * p[1] + p[2];
+	float64_t2 B = 2.0 * (p[1] - p[0]);
+	float64_t2 C = p[0];
+	
+	auto xroots = hatchutils::getCurveRoot(A.x, B.x, C.x);
+	auto yroots = hatchutils::getCurveRoot(A.y, B.y, C.y);
+
+	return { xroots.x, xroots.y, yroots.x, yroots.y };
+}
+
+Hatch::QuadraticBezier Hatch::QuadraticBezier::splitCurveTakeLeft(double t) const
+{
+	QuadraticBezier outputCurve;
+	outputCurve.p[0] = p[0];
+	outputCurve.p[1] = (1 - t) * p[0] + t * p[1];
+	outputCurve.p[2] = (1 - t) * ((1 - t) * p[0] + t * p[1]) + t * ((1 - t) * p[1] + t * p[2]);
+	auto p0 = outputCurve.p[0];
+	auto p1 = outputCurve.p[1];
+	auto p2 = outputCurve.p[2];
+
+	return outputCurve;
+}
+
+Hatch::QuadraticBezier Hatch::QuadraticBezier::splitCurveTakeRight(double t) const
+{
+	QuadraticBezier outputCurve;
+	outputCurve.p[0] = p[2];
+	outputCurve.p[1] = (1 - t) * p[1] + t * p[2];
+	outputCurve.p[2] = (1 - t) * ((1 - t) * p[0] + t * p[1]) + t * ((1 - t) * p[1] + t * p[2]);
+	auto p0 = outputCurve.p[0];
+	auto p1 = outputCurve.p[1];
+	auto p2 = outputCurve.p[2];
+
+	return outputCurve;
+}
+
+std::array<Hatch::QuadraticBezier, 2> Hatch::QuadraticBezier::splitIntoMonotonicSegments(int major) const
+{
+	// Getting derivatives for our quadratic bezier
+	auto a = 2.0 * (index(p[1], major) - index(p[0], major));
+	auto b = 2.0 * (index(p[2], major) - index(p[1], major));
+
+	// Finding roots for the quadratic bezier derivatives (a straight line)
+	auto rcp = 1.0 / (b - a);
+	auto t = -a * rcp;
+	if (isinf(rcp) || t <= 0.0 || t >= 1.0) return { { float64_t2(nbl::core::nan<double>()) }};
```

Comment:
this is not at all a clean way to return an invalid function. 
simply take return type as a reference argument and return a `bool`